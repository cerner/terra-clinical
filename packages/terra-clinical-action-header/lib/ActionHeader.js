'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _terraButton = require('terra-button');

var _terraButton2 = _interopRequireDefault(_terraButton);

var _terraButtonGroup = require('terra-button-group');

var _terraButtonGroup2 = _interopRequireDefault(_terraButtonGroup);

var _IconLeft = require('terra-icon/lib/icon/IconLeft');

var _IconLeft2 = _interopRequireDefault(_IconLeft);

var _IconClose = require('terra-icon/lib/icon/IconClose');

var _IconClose2 = _interopRequireDefault(_IconClose);

var _IconMinimize = require('terra-icon/lib/icon/IconMinimize');

var _IconMinimize2 = _interopRequireDefault(_IconMinimize);

var _IconMaximize = require('terra-icon/lib/icon/IconMaximize');

var _IconMaximize2 = _interopRequireDefault(_IconMaximize);

var _IconChevronUp = require('terra-icon/lib/icon/IconChevronUp');

var _IconChevronUp2 = _interopRequireDefault(_IconChevronUp);

var _IconChevronDown = require('terra-icon/lib/icon/IconChevronDown');

var _IconChevronDown2 = _interopRequireDefault(_IconChevronDown);

var _terraClinicalHeader = require('terra-clinical-header');

var _terraClinicalHeader2 = _interopRequireDefault(_terraClinicalHeader);

var _terraResponsiveElement = require('terra-responsive-element');

var _terraResponsiveElement2 = _interopRequireDefault(_terraResponsiveElement);

require('terra-base/lib/baseStyles');

require('./ActionHeader.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  /**
   * Text to be displayed as the title in the header bar
   **/
  title: _propTypes2.default.string,

  /**
   * Callback function for when the close button is clicked. The back button will not display if this is not set.
   * On small viewports a back button will be displayed instead of a close button when a separate onBack callback is not set.
   **/
  onClose: _propTypes2.default.func,

  /**
   * Callback function for when the back button is clicked. The back button will not display if this is not set.
   **/
  onBack: _propTypes2.default.func,

  /**
   * Callback function for when the expand button is clicked.
   * The expand button will not display if this is not set or on small viewports.
   * Only the expand button will be rendered if onMaximize and onMinimize are set.
   **/
  onMaximize: _propTypes2.default.func,

  /**
   * Callback function for when the minimize button is clicked.
   * The minimize button will not display if this is not set or on small viewports.
   * Only the expand button will be rendered if both onMaximize and onMinimize are set.
   **/
  onMinimize: _propTypes2.default.func,

  /**
   * Callback function for when the next button is clicked. The previous-next button group will not display if neither this or onPrevious are set.
   **/
  onNext: _propTypes2.default.func,

  /**
   * Callback function for when the previous button is clicked. The previous-next button group will not display if neither this or onNext are set.
   **/
  onPrevious: _propTypes2.default.func,

  /**
   * Child element to be displayed on the right end of the header.
   **/
  children: _propTypes2.default.element
};

var defaultProps = {
  title: '',
  onClose: null,
  onBack: null,
  onMaximize: null,
  onMinimize: null,
  onNext: null,
  onPrevious: null,
  children: null
};

var ActionHeader = function ActionHeader(_ref) {
  var title = _ref.title,
      onClose = _ref.onClose,
      onBack = _ref.onBack,
      onMaximize = _ref.onMaximize,
      onMinimize = _ref.onMinimize,
      onPrevious = _ref.onPrevious,
      onNext = _ref.onNext,
      children = _ref.children,
      customProps = _objectWithoutProperties(_ref, ['title', 'onClose', 'onBack', 'onMaximize', 'onMinimize', 'onPrevious', 'onNext', 'children']);

  var attributes = _extends({}, customProps);
  var closeButton = onClose ? _react2.default.createElement(_terraButton2.default, { icon: _react2.default.createElement(_IconClose2.default, null), onClick: onClose }) : null;
  var backButton = onBack ? _react2.default.createElement(_terraButton2.default, { icon: _react2.default.createElement(_IconLeft2.default, null), onClick: onBack }) : null;

  var closeButtonSmall = void 0;
  var backButtonSmall = void 0;
  if (onClose && !onBack) {
    backButtonSmall = _react2.default.createElement(_terraButton2.default, { icon: _react2.default.createElement(_IconLeft2.default, null), onClick: onClose });
    closeButtonSmall = null;
  } else {
    closeButtonSmall = closeButton;
    backButtonSmall = backButton;
  }

  var expandButton = void 0;
  if (!backButton) {
    if (onMaximize) {
      expandButton = _react2.default.createElement(_terraButton2.default, { icon: _react2.default.createElement(_IconMaximize2.default, null), onClick: onMaximize });
    } else if (onMinimize) {
      expandButton = _react2.default.createElement(_terraButton2.default, { icon: _react2.default.createElement(_IconMinimize2.default, null), onClick: onMinimize });
    }
  }

  var previousNextButtonGroup = null;
  if (onPrevious || onNext) {
    previousNextButtonGroup = _react2.default.createElement(
      _terraButtonGroup2.default,
      null,
      _react2.default.createElement(_terraButtonGroup2.default.Button, { icon: _react2.default.createElement(_IconChevronUp2.default, null), onClick: onPrevious, key: 'ActionHeaderPrevious' }),
      _react2.default.createElement(_terraButtonGroup2.default.Button, { icon: _react2.default.createElement(_IconChevronDown2.default, null), onClick: onNext, key: 'ActionHeaderNext' })
    );
  }

  var leftButtonsDefault = _react2.default.createElement(
    'div',
    { className: 'terraClinical-ActionHeader-leftButtons' },
    backButton,
    expandButton,
    previousNextButtonGroup
  );

  var rightButtonsDefault = _react2.default.createElement(
    'div',
    { className: 'terraClinical-ActionHeader-rightButtons' },
    children,
    closeButton
  );

  var leftButtonsSmall = _react2.default.createElement(
    'div',
    { className: 'terraClinical-ActionHeader-leftButtons' },
    backButtonSmall,
    previousNextButtonGroup
  );

  var rightButtonsSmall = _react2.default.createElement(
    'div',
    { className: 'terraClinical-ActionHeader-rightButtons' },
    children,
    closeButtonSmall
  );

  var actionHeader = _react2.default.createElement(_terraClinicalHeader2.default, _extends({}, attributes, {
    startContent: leftButtonsDefault,
    title: title,
    endContent: rightButtonsDefault
  }));

  var smallActionHeader = _react2.default.createElement(_terraClinicalHeader2.default, _extends({}, attributes, {
    startContent: leftButtonsSmall,
    title: title,
    endContent: rightButtonsSmall
  }));

  return _react2.default.createElement(_terraResponsiveElement2.default, {
    responsiveTo: 'window',
    defaultElement: smallActionHeader,
    small: actionHeader
  });
};

ActionHeader.propTypes = propTypes;
ActionHeader.defaultProps = defaultProps;

exports.default = ActionHeader;