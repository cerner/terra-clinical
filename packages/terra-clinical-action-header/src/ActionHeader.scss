@import '~terra-mixins';

.terraClinical-Header .terraClinical-ActionHeader-rightButtons > * {
  @include terra-margin-start(1.429rem);
  @include terra-margin-end(0);

  margin-bottom: 0;
  margin-top: 0;
}

.terraClinical-Header .terraClinical-ActionHeader-leftButtons > * {
  @include terra-margin-start(0);
  @include terra-margin-end(1.429rem);

  margin-bottom: 0;
  margin-top: 0;
}
