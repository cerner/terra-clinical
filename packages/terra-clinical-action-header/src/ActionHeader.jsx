import React from 'react';
import classNames from 'classnames/bind';
import PropTypes from 'prop-types';
import Button from 'terra-button';
import ButtonGroup from 'terra-button-group';
import IconLeft from 'terra-icon/lib/icon/IconLeft';
import IconClose from 'terra-icon/lib/icon/IconClose';
import IconMinimize from 'terra-icon/lib/icon/IconMinimize';
import IconMaximize from 'terra-icon/lib/icon/IconMaximize';
import IconChevronUp from 'terra-icon/lib/icon/IconChevronUp';
import IconChevronDown from 'terra-icon/lib/icon/IconChevronDown';
import Header from 'terra-clinical-header';
import ResponsiveElement from 'terra-responsive-element';
import 'terra-base/lib/baseStyles';
import styles from './ActionHeader.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Text to be displayed as the title in the header bar
   */
  title: PropTypes.string,

  /**
   * Callback function for when the close button is clicked. The back button will not display if this is not set.
   * On small viewports a back button will be displayed instead of a close button when a separate onBack callback is not set.
   */
  onClose: PropTypes.func,

  /**
   * Callback function for when the back button is clicked. The back button will not display if this is not set.
   */
  onBack: PropTypes.func,

  /**
   * Callback function for when the expand button is clicked.
   * The expand button will not display if this is not set or on small viewports.
   * Only the expand button will be rendered if onMaximize and onMinimize are set.
   */
  onMaximize: PropTypes.func,

  /**
   * Callback function for when the minimize button is clicked.
   * The minimize button will not display if this is not set or on small viewports.
   * Only the expand button will be rendered if both onMaximize and onMinimize are set.
   */
  onMinimize: PropTypes.func,

  /**
   * Callback function for when the next button is clicked. The previous-next button group will not display if neither this or onPrevious are set.
   */
  onNext: PropTypes.func,

  /**
   * Callback function for when the previous button is clicked. The previous-next button group will not display if neither this or onNext are set.
   */
  onPrevious: PropTypes.func,

  /**
   * Child element to be displayed on the right end of the header.
   */
  children: PropTypes.element,
};

const defaultProps = {
  title: '',
  onClose: null,
  onBack: null,
  onMaximize: null,
  onMinimize: null,
  onNext: null,
  onPrevious: null,
  children: null,
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Please add locale prop to Base component to load translations');
    }
  },
};

const ActionHeader = ({
  title,
  onClose,
  onBack,
  onMaximize,
  onMinimize,
  onPrevious,
  onNext,
  children,
  ...customProps }, {
  intl,
}) => {
  const attributes = Object.assign({}, customProps);
  const backText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.back' });
  const closeText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.close' });
  const minimizeText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.minimize' });
  const maximizeText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.maximize' });
  const previousText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.previous' });
  const nextText = intl.formatMessage({ id: 'Terra.Clinical.ActionHeader.next' });

  const closeButton = onClose ? <Button icon={<IconClose />} aria-label={closeText} onClick={onClose} /> : null;
  const backButton = onBack ? <Button icon={<IconLeft />} aria-label={backText} onClick={onBack} /> : null;

  let closeButtonSmall;
  let backButtonSmall;
  if (onClose && !onBack) {
    backButtonSmall = <Button icon={<IconLeft />} aria-label={backText} onClick={onClose} />;
    closeButtonSmall = null;
  } else {
    closeButtonSmall = closeButton;
    backButtonSmall = backButton;
  }


  let expandButton;
  if (!backButton) {
    if (onMaximize) {
      expandButton = <Button icon={<IconMaximize />} aria-label={maximizeText} onClick={onMaximize} />;
    } else if (onMinimize) {
      expandButton = <Button icon={<IconMinimize />} aria-label={minimizeText} onClick={onMinimize} />;
    }
  }

  let previousNextButtonGroup = null;
  if (onPrevious || onNext) {
    previousNextButtonGroup = (
      <ButtonGroup>
        <ButtonGroup.Button icon={<IconChevronUp />} aria-label={previousText} onClick={onPrevious} key="ActionHeaderPrevious" />
        <ButtonGroup.Button icon={<IconChevronDown />} aria-label={nextText} onClick={onNext} key="ActionHeaderNext" />
      </ButtonGroup>
    );
  }

  const leftButtonsDefault = (
    <div className={cx('left-buttons')}>
      {backButton}
      {expandButton}
      {previousNextButtonGroup}
    </div>
  );

  const rightButtonsDefault = (
    <div className={cx('right-buttons')}>
      {children}
      {closeButton}
    </div>
  );

  const leftButtonsSmall = (
    <div className={cx('left-buttons')}>
      {backButtonSmall}
      {previousNextButtonGroup}
    </div>
  );

  const rightButtonsSmall = (
    <div className={cx('right-buttons')}>
      {children}
      {closeButtonSmall}
    </div>
  );

  const actionHeader = (
    <Header
      {...attributes}
      data-terra-clincial-action-header
      startContent={leftButtonsDefault}
      title={title}
      endContent={rightButtonsDefault}
    />
  );

  const smallActionHeader = (
    <Header
      {...attributes}
      data-terra-clincial-action-header
      startContent={leftButtonsSmall}
      title={title}
      endContent={rightButtonsSmall}
    />
  );

  return (
    <ResponsiveElement
      responsiveTo="window"
      defaultElement={smallActionHeader}
      small={actionHeader}
    />
  );
};

ActionHeader.propTypes = propTypes;
ActionHeader.defaultProps = defaultProps;
ActionHeader.contextTypes = contextTypes;

export default ActionHeader;
