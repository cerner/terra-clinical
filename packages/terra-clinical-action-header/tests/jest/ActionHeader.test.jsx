import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { IntlProvider } from 'react-intl';
import Button from 'terra-button';
import ActionHeader from '../../src/ActionHeader';
import messages from '../../translations/en-US.json';

const locale = 'en-US';

// Snapshot tests

it('should render default component', () => {
  const defaultHeader = <IntlProvider locale={locale} messages={messages}><ActionHeader /></IntlProvider>;
  const wrapper = shallow(defaultHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with title', () => {
  const titleActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader title="Title" />
    </IntlProvider>
  );
  const wrapper = shallow(titleActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with close button', () => {
  const closeActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onClose={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(closeActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with back button', () => {
  const backActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onBack={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(backActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with expand button', () => {
  const expandActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onMaximize={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(expandActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with minimize button', () => {
  const minimizeActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onMinimize={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(minimizeActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with Previous-Next button when onPrevious is givien', () => {
  const minimizeActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onPrevious={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(minimizeActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with Previous-Next button when onNext is givien', () => {
  const minimizeActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader onNext={() => 1} />
    </IntlProvider>
  );
  const wrapper = shallow(minimizeActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('should render ActionHeader with children', () => {
  const minimizeActionHeader = (
    <IntlProvider locale={locale} messages={messages}>
      <ActionHeader>
        <Button />
      </ActionHeader>
    </IntlProvider>
  );
  const wrapper = shallow(minimizeActionHeader);
  expect(wrapper).toMatchSnapshot();
});

it('throws error on missing locale prop in Base', () => {
  try {
    shallow(<ActionHeader />);
  } catch (e) {
    expect(e.message).toContain('add locale prop to Base component');
  }
});
