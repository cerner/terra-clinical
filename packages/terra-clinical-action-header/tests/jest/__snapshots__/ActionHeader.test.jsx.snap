// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ActionHeader with Previous-Next button when onNext is givien 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={[Function]}
  onPrevious={null}
  title=""
/>
`;

exports[`should render ActionHeader with Previous-Next button when onPrevious is givien 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={[Function]}
  title=""
/>
`;

exports[`should render ActionHeader with back button 1`] = `
<ActionHeader
  onBack={[Function]}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title=""
/>
`;

exports[`should render ActionHeader with children 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title=""
>
  <Button
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isReversed={false}
    type="button"
    variant="default"
  />
</ActionHeader>
`;

exports[`should render ActionHeader with close button 1`] = `
<ActionHeader
  onBack={null}
  onClose={[Function]}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title=""
/>
`;

exports[`should render ActionHeader with expand button 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={[Function]}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title=""
/>
`;

exports[`should render ActionHeader with minimize button 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={[Function]}
  onNext={null}
  onPrevious={null}
  title=""
/>
`;

exports[`should render ActionHeader with title 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title="Title"
/>
`;

exports[`should render default component 1`] = `
<ActionHeader
  onBack={null}
  onClose={null}
  onMaximize={null}
  onMinimize={null}
  onNext={null}
  onPrevious={null}
  title=""
/>
`;
